docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,289 @@
+# ADR 015: IBC Packet Receiver
+
+## Changelog
+
+- 2019 Oct 22: Initial Draft
+
+## Context
+ 
+[ICS 26 - Routing Module](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module) defines a function [`handlePacketRecv`](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module#packet-relay).
+
+In ICS 26, the routing module is defined as a layer above each application module
+which verifies and routes messages to the destination modules. It is possible to 
+implement it as a separate module, however, we already have functionality to route
+messages upon the destination identifiers in the baseapp. This ADR suggests 
+to utilize existing `baseapp.router` tp route packets to application modules.
+
+Generally, each routing module callbacks have two separate steps in them, 
+verificaton and execution. This corresponds to the `AnteHandler`-`Handler`
+moodel inside the SDK. We can do the verificaton inside the `AnteHandler`
+in order to increase developer ergonomics by reducing boilerplate 
+verification code.
+
+## Decision
+
+`PortKeeper` will have the capability key that is able to access only on the 
+channels binded to the port. Entities those holding a `PortKeeper` will be
+able to call the corresponding `ChannelKeeper`s method, but only with the
+allowed port. `ChannelKeeper.Port(string, ChannelChecker)` will be defined to
+easily construct a capability-safe `PortKeeper`. This will be address in 
+another ADR and we will use unsecure `ChannelKeeper` for now.
+
+`baseapp.runMsgs` will break the loop over the messages if one of the handlers
+returns `!Result.IsOK()`. However, the outer logic will write the cached 
+store if `Result.IsOK() || Result.Code.IsBreak()`. `Result.Code.IsBreak()` if
+`Result.Code == CodeTxBreak`.
```

Comment:
Could you explain why we might want to write the cached logic sometimes even if one of the messages fails?

Reply:
Because we want to persist sequence number increments & commitments even if IBC packet execution fails. It's basically analogous to nonces on accounts incrementing even if the tx fails in `DeliverTx`.

Reply:
> Because we want to persist sequence number increments & commitments even if IBC packet execution fails. It's basically analogous to nonces on accounts incrementing even if the tx fails in DeliverTx.

this needs to be written down